File: FSUIKit/Classes/FSAutoLayoutButtonsView.m
//
//  FSAutoLayoutButtonsView.m
//  myhome
//
//  Created by FudonFuchina on 2018/8/18.
//  Copyright © 2018年 fuhope. All rights reserved.
//

#import "FSAutoLayoutButtonsView.h"
#import "FSCalculator.h"

#define _key_value_button_tag 1000

@implementation FSAutoLayoutButtonsView{
    CGFloat _height;
}

- (void)setTexts:(NSArray<NSString *> *)texts{
    _texts = nil;
    if ([texts isKindOfClass:NSArray.class] && texts.count) {
        _texts = texts;
    }
    
    Class _class_UIButton = UIButton.class;
    for (UIButton *sub in self.subviews) {
        if ([sub isKindOfClass:_class_UIButton] && sub.tag > _key_value_button_tag) {
            sub.hidden = YES;
            [sub removeFromSuperview];
        }
    }
    
    Class _class_NSString = NSString.class;
    UIFont *font = [UIFont systemFontOfSize:13 weight:UIFontWeightMedium];
    CGFloat sw = UIScreen.mainScreen.bounds.size.width;
    NSInteger lrMargin = 15;
    CGFloat rightMargin = sw - lrMargin;
    CGFloat offsetX = lrMargin;
    CGFloat offsetY = 20;
    CGFloat bh = 44;
    NSInteger xSpace = 30;
    for (int x = 0; x < texts.count; x ++) {
        NSString *item = texts[x];
        if (![item isKindOfClass:_class_NSString]) {
            continue;
        }
        
        CGFloat w = [FSCalculator textWidth:item font:font labelHeight:30] + 40;
        if ((offsetX + w) > rightMargin) {
            offsetY += 70;
            offsetX = lrMargin;
        }
        
        UIButton *button = [UIButton buttonWithType:UIButtonTypeSystem];
        button.frame = CGRectMake(offsetX, offsetY, w, bh);
        button.tag = x + _key_value_button_tag + 1;
        button.backgroundColor = UIColor.whiteColor;
        [button setTitle:item forState:UIControlStateNormal];
        [button addTarget:self action:@selector(btnClick:) forControlEvents:UIControlEventTouchUpInside];
        [self addSubview:button];
        
        offsetX += (w + xSpace);
        
        if (self.configButton) {
            self.configButton(button);
        }
    }
    _height = offsetY + bh;
}

- (CGFloat)selfHeight{
    return  _height;
}

- (void)btnClick:(UIButton *)button{
    if (self.click) {
        NSInteger index = button.tag - _key_value_button_tag - 1;
        self.click(self, index);
    }
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end

File: FSUIKit/Classes/Swift/FSAutoLayoutButtonsView.swift
// FSAutoLayoutButtonsViewS.swift
// Translated from FSAutoLayoutButtonsView.h/m

import UIKit

public class FSAutoLayoutButtonsView: UIView {
    
    private let keyValueButtonTag = 1000
    private var _height: CGFloat = 0
    
    public var texts: [String]? {
        didSet {
            setupButtons()
        }
    }
    
    public var click: ((FSAutoLayoutButtonsView, Int) -> Void)?
    public var configButton: ((UIButton) -> Void)?
    
    public var selfHeight: CGFloat {
        return _height
    }
    
    private func setupButtons() {
        // Remove existing buttons
        for sub in subviews {
            if let button = sub as? UIButton, button.tag > keyValueButtonTag {
                button.isHidden = true
                button.removeFromSuperview()
            }
        }
        
        guard let texts = texts, !texts.isEmpty else { return }
        
        let font = UIFont.systemFont(ofSize: 13, weight: .medium)
        let sw = UIScreen.main.bounds.width
        let lrMargin: CGFloat = 15
        let rightMargin = sw - lrMargin
        var offsetX: CGFloat = lrMargin
        var offsetY: CGFloat = 20
        let bh: CGFloat = 44
        let xSpace: CGFloat = 30
        
        for (index, item) in texts.enumerated() {
            let w = textWidth(item, font: font, labelHeight: 30) + 40
            if (offsetX + w) > rightMargin {
                offsetY += 70
                offsetX = lrMargin
            }
            
            let button = UIButton(type: .system)
            button.frame = CGRect(x: offsetX, y: offsetY, width: w, height: bh)
            button.tag = index + keyValueButtonTag + 1
            button.backgroundColor = .white
            button.setTitle(item, for: .normal)
            button.addTarget(self, action: #selector(btnClick(_:)), for: .touchUpInside)
            addSubview(button)
            
            offsetX += (w + xSpace)
            
            configButton?(button)
        }
        _height = offsetY + bh
    }
    
    @objc private func btnClick(_ button: UIButton) {
        let index = button.tag - keyValueButtonTag - 1
        click?(self, index)
    }
    
    private func textWidth(_ text: String, font: UIFont, labelHeight: CGFloat) -> CGFloat {
        let size = CGSize(width: CGFloat.greatestFiniteMagnitude, height: labelHeight)
        let attributes: [NSAttributedString.Key: Any] = [.font: font]
        let rect = (text as NSString).boundingRect(with: size, options: .usesLineFragmentOrigin, attributes: attributes, context: nil)
        return ceil(rect.width)
    }
}
